File: services/ai-service/src/routes/aiRoutes.js
const { Router } = require('express');
const { chatWithTools } = require('../services/deepseekService');
const { buildSystemPrompt } = require('../services/systemPrompt');
const { buildSystemContext } = require('../services/contextBuilder');
const { TOOLS, executeTool } = require('../services/toolExecutor');
const { generateInsights } = require('../services/insightsService');

const router = Router();

/**
 * POST /api/ai/ask
 * Body: { message: string, conversation?: Array<{role,content}> }
 */
router.post('/ask', async (req, res) => {
  const { message, conversation } = req.body || {};

  if (!message || typeof message !== 'string' || !message.trim()) {
    return res.status(400).json({ error: 'El mensaje es requerido' });
  }

  try {
    const contextData = await buildSystemContext().catch(() => null);
    const systemPrompt = buildSystemPrompt(contextData);
    const systemMsg = { role: 'system', content: systemPrompt };

    const history = Array.isArray(conversation) ? conversation.slice(-20) : [];
    const messages = [systemMsg, ...history, { role: 'user', content: message }];

    const result = await chatWithTools(messages, TOOLS, executeTool);

    res.json({
      response: result.response,
      usage: result.usage
        ? {
            input_tokens: result.usage.prompt_tokens,
            output_tokens: result.usage.completion_tokens,
            total_tokens: result.usage.total_tokens,
          }
        : null,
    });
  } catch (err) {
    console.error('[AI Service] Error:', err.message);
    res.status(500).json({
      error: 'Error al procesar la consulta',
      detail: process.env.NODE_ENV === 'development' ? err.message : undefined,
    });
  }
});

/**
 * GET /api/ai/insights
 * Genera insights automáticos del negocio (con caché de 5 min)
 */
router.get('/insights', async (_req, res) => {
  try {
    const insights = await generateInsights();
    res.json(insights);
  } catch (err) {
    console.error('[AI Service] Insights error:', err.message);
    res.status(503).json({
      error: 'No se pudieron generar insights en este momento',
      insight: 'Estamos teniendo problemas para analizar los datos. Intenta de nuevo más tarde.',
      trend_percentage: 0,
      trend_direction: 'up',
      trend_comparison: 'vs semana pasada',
      timestamp: new Date().toISOString(),
    });
  }
});

/**
 * POST /api/ai/insights/refresh
 * Fuerza la regeneración de insights (limpia caché)
 */
router.post('/insights/refresh', async (_req, res) => {
  try {
    const { clearCache } = require('../services/insightsService');
    clearCache();
    const insights = await generateInsights();
    res.json(insights);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

/**
 * GET /api/ai/health
 */
router.get('/health', (_req, res) => {
  res.json({ status: 'ok', service: 'ai-service' });
});

/**
 * POST /api/ai/telegram-webhook
 */
router.post('/telegram-webhook', async (req, res) => {
  const { text, chatId } = req.body;
  if (!text || !chatId) return res.status(400).json({ error: 'Missing text or chatId' });

  const { saveTelegramHistory, getTelegramHistory, sendTelegramNotification } = require('../services/telegramService');

  try {
    // 1. Guardar mensaje del usuario en el historial
    await saveTelegramHistory(chatId, 'user', text);

    // 2. Recuperar historial
    const history = await getTelegramHistory(chatId);

    // 3. Obtener contexto
    const contextData = await buildSystemContext().catch(() => null);
    const systemPrompt = buildSystemPrompt(contextData);
    const systemMsg = { role: 'system', content: systemPrompt };

    const messages = [systemMsg, ...history];

    // 4. Preguntar a la IA
    const result = await chatWithTools(messages, TOOLS, executeTool);

    // 5. Guardar respuesta en historial
    await saveTelegramHistory(chatId, 'assistant', result.response);

    // 6. Enviar notificación a Telegram (vía Redis)
    await sendTelegramNotification('🤖 Kiora AI', result.response);

    res.json({ success: true });
  } catch (err) {
    console.error('[AI Webhook] Error:', err.message);
    res.status(500).json({ error: err.message });
  }
});

/**
 * POST /api/ai/close-business
 */
router.post('/close-business', async (req, res) => {
  const { sendTelegramNotification, redisClient } = require('../services/telegramService');
  try {
    // 1. Bloquear negocio en Redis para no permitir más ventas (Orders Service lo debería leer)
    await redisClient.set('kiora:business_state', 'closed');

    // 2. Pedirle a la IA que genere un resumen del día
    const contextData = await buildSystemContext().catch(() => null);
    const systemPrompt = buildSystemPrompt(contextData);
    const systemMsg = { role: 'system', content: systemPrompt };
    const promptMsg = { 
      role: 'user', 
      content: 'El negocio acaba de ser cerrado por el administrador. Genera un resumen operativo del día (Ventas de hoy, productos más vendidos, stock bajo). Hazlo en un formato agradable para enviarlo por Telegram. ¡Es tu reporte de cierre diario!' 
    };

    const messages = [systemMsg, promptMsg];
    const result = await chatWithTools(messages, TOOLS, executeTool);

    // 3. Enviar a Telegram
    await sendTelegramNotification('🛑 Cierre de Negocio', result.response);

    res.json({ success: true, message: 'Negocio cerrado. Resumen enviado por Telegram.' });
  } catch (err) {
    console.error('[AI Close Business] Error:', err.message);
    res.status(500).json({ error: err.message });
  }
});

/**
 * POST /api/ai/open-business
 */
router.post('/open-business', async (req, res) => {
  const { sendTelegramNotification, redisClient } = require('../services/telegramService');
  try {
    await redisClient.set('kiora:business_state', 'open');
    await sendTelegramNotification('🟢 Apertura de Negocio', 'El negocio ha sido abierto y el sistema está listo para recibir ventas.');
    res.json({ success: true, message: 'Negocio abierto.' });
  } catch (err) {
    console.error('[AI Open Business] Error:', err.message);
    res.status(500).json({ error: err.message });
  }
});

/**
 * GET /api/ai/business-state
 */
router.get('/business-state', async (req, res) => {
  const { redisClient } = require('../services/telegramService');
  try {
    const state = await redisClient.get('kiora:business_state');
    // Si no existe, asumimos que está abierto
    res.json({ state: state || 'open' });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;

File: services/ai-service/app/routes/ai_routes.py
from fastapi import APIRouter, HTTPException, Body, Request
from typing import List, Optional
from datetime import datetime
import time
import logging

from fastapi.responses import StreamingResponse
from app.services.deepseek_service import chat_with_tools, stream_chat_with_tools
from app.services.system_prompt import build_system_prompt
from app.services.context_builder import build_system_context
from app.services.insights_service import generate_insights, clear_cache
from app.services.telegram_service import save_telegram_history, get_telegram_history, send_telegram_notification, redis_client

from app.services.tool_executor import fetch_json, GATEWAY
from app.services.ml_service import analyze_basket_cross_selling_ml

logger = logging.getLogger(__name__)

router = APIRouter()

_cross_selling_cache = None
_cross_selling_expiry = 0
CROSS_SELLING_TTL = 3600

@router.post("/cross-selling/recommend")
async def recommend_cross_selling(payload: dict = Body(...)):
    """Endpoint para el Front del Cliente: Analiza el carrito y sugiere productos estratégicos."""
    global _cross_selling_cache, _cross_selling_expiry
    cart_items = payload.get("cart_items", [])
    
    if not isinstance(cart_items, list):
        raise HTTPException(status_code=400, detail="cart_items debe ser una lista de IDs de productos")
        
    now = time.time()
    if not _cross_selling_cache or now >= _cross_selling_expiry:
        orders = await fetch_json(f"{GATEWAY}/orders?limit=300")
        products = await fetch_json(f"{GATEWAY}/products")
        if isinstance(orders, dict) and "error" in orders:
            orders = []
        if isinstance(products, dict) and "error" in products:
            products = []
            
        _cross_selling_cache = analyze_basket_cross_selling_ml(orders, products)
        _cross_selling_expiry = now + CROSS_SELLING_TTL

    recommendations = []
    if _cross_selling_cache and "top_cross_selling_opportunities" in _cross_selling_cache:
        # Analizamos las reglas ML contra el carrito del cliente
        for opp in _cross_selling_cache["top_cross_selling_opportunities"]:
            id_A = opp["product_id_A"]
            id_B = opp["product_id_B"]
            
            # Si tiene A pero NO B, sugerir B
            if id_A in cart_items and id_B not in cart_items:
                recommendations.append({"product_id": id_B, "name": opp["product_B"], "confidence": opp["confidence_percent"]})
            # Si tiene B pero NO A, sugerir A
            elif id_B in cart_items and id_A not in cart_items:
                recommendations.append({"product_id": id_A, "name": opp["product_A"], "confidence": opp["confidence_percent"]})
                
    # Ordenar por mayor confianza
    recommendations.sort(key=lambda x: x["confidence"], reverse=True)
    
    # Filtrar duplicados tomando los mejores 3
    seen = set()
    final_recs = []
    for r in recommendations:
        if r["product_id"] not in seen:
            seen.add(r["product_id"])
            final_recs.append(r)
            if len(final_recs) >= 3:
                break
                
    return {"recommendations": final_recs}


@router.post("/ask")
async def ask(payload: dict = Body(...)):
    message = payload.get("message")
    conversation = payload.get("conversation", [])

    if not message or not isinstance(message, str) or not message.strip():
        raise HTTPException(status_code=400, detail="El mensaje es requerido")

    try:
        try:
            context_data = await build_system_context()
        except Exception:
            context_data = None
            
        system_prompt = build_system_prompt(context_data)
        system_msg = {"role": "system", "content": system_prompt}

        history = conversation[-20:] if isinstance(conversation, list) else []
        messages = [system_msg] + history + [{"role": "user", "content": message}]

        # Return a streaming response
        return StreamingResponse(
            stream_chat_with_tools(messages),
            media_type="text/event-stream"
        )

    except Exception as e:
        logger.error(f"[AI Service] Error: {e}")
        raise HTTPException(status_code=500, detail="Error al procesar la consulta")

@router.get("/insights")
async def get_insights():
    try:
        insights = await generate_insights()
        return insights
    except Exception as e:
        logger.error(f"[AI Service] Insights error: {e}")
        raise HTTPException(status_code=503, detail={
            "error": "No se pudieron generar insights en este momento",
            "insight": "Estamos teniendo problemas para analizar los datos. Intenta de nuevo más tarde.",
            "trend_percentage": 0,
            "trend_direction": "up",
            "trend_comparison": "vs semana pasada",
            "timestamp": datetime.utcnow().isoformat() + "Z"
        })

@router.post("/insights/refresh")
async def refresh_insights():
    try:
        clear_cache()
        insights = await generate_insights()
        return insights
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/health")
async def health():
    return {"status": "ok", "service": "ai-service"}

@router.post("/telegram-webhook")
async def telegram_webhook(payload: dict = Body(...)):
    text = payload.get("text")
    chat_id = payload.get("chatId")
    
    if not text or not chat_id:
        raise HTTPException(status_code=400, detail="Missing text or chatId")

    try:
        await save_telegram_history(chat_id, "user", text)
        history = await get_telegram_history(chat_id)

        try:
            context_data = await build_system_context()
        except Exception:
            context_data = None
            
        system_prompt = build_system_prompt(context_data)
        system_msg = {"role": "system", "content": system_prompt}

        messages = [system_msg] + history
        result = await chat_with_tools(messages)

        await save_telegram_history(chat_id, "assistant", result["response"])
        await send_telegram_notification('🤖 Kiora AI', result["response"])

        return {"success": True}
    except Exception as e:
        logger.error(f"[AI Webhook] Error: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/close-business")
async def close_business():
    try:
        await redis_client.set("kiora:business_state", "closed")

        try:
            context_data = await build_system_context()
        except Exception:
            context_data = None
            
        system_prompt = build_system_prompt(context_data)
        system_msg = {"role": "system", "content": system_prompt}
        
        prompt_msg = {
            "role": "user",
            "content": "El negocio acaba de ser cerrado por el administrador. Genera un resumen operativo del día (Ventas de hoy, productos más vendidos, stock bajo). Hazlo en un formato agradable para enviarlo por Telegram. ¡Es tu reporte de cierre diario!"
        }

        messages = [system_msg, prompt_msg]
        result = await chat_with_tools(messages)

        await send_telegram_notification('🛑 Cierre de Negocio', result["response"])

        return {"success": True, "message": "Negocio cerrado. Resumen enviado por Telegram."}
    except Exception as e:
        logger.error(f"[AI Close Business] Error: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/open-business")
async def open_business():
    try:
        await redis_client.set("kiora:business_state", "open")
        await send_telegram_notification('🟢 Apertura de Negocio', 'El negocio ha sido abierto y el sistema está listo para recibir ventas.')
        return {"success": True, "message": "Negocio abierto."}
    except Exception as e:
        logger.error(f"[AI Open Business] Error: {e}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/business-state")
async def get_business_state():
    try:
        state = await redis_client.get("kiora:business_state")
        return {"state": state if state else "open"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
